src/ucp/core/ucp_request.inl

```diff
@@ -663,15 +663,26 @@ ucp_request_recv_data_unpack(ucp_request_t *req, const void *data,
 }
 
 static UCS_F_ALWAYS_INLINE ucp_recv_desc_t*
-ucp_recv_desc_get(ucp_worker_h worker, size_t length)
+ucp_recv_desc_alloc(ucp_worker_h worker, size_t length)
 {
-    /* TODO: Optimize */
-    uint32_t rp  = ucs_roundup_pow2(length + 1); /* make sure length is not 0 */
-    unsigned idx = ucs_popcount(worker->am_mps_map & (rp - 1));
+    size_t size = length + 1;
+    unsigned idx;
 
-    ucs_assertv(length < UINT_MAX, "length %zu", length);
+    /* - use size equal to length + 1 to avoid passing 0 to
+     *   ucs_count_leading_zero_bits()
+     * - do not roundup the size if it is pow of 2 already, to not use bigger
+     *   mpool than needed
+     */
+    idx = (sizeof(size_t) * 8) -
```

Comment:
can we use UCP_WORKER_AM_MPS_MAP_SIZE here?
AFAIU the definition of UCP_WORKER_AM_MPS_MAP_SIZE means that length cannot be larger than 32bit
so maybe do something like
UCP_WORKER_AM_MPS_MAP_SIZE  - (ucs_count_leading_zero_bits32(length - 1) + 1)
or even arrange the mpools in the array in such way that can always do something like
`idx = ucs_count_leading_zero_bits32(length - 1)`